### ISSUE TITLE
G722 codec change issue

### ISSUE THREAD

**Issue by melekoktay** [signalwire/freeswitch#1181]
Hi,

We found interesting issue when working on FreeSwitch on the Field.  FreeSwitch selects **G722** codec for call and during the call, if codec changes for somehow (like hold action or codec change request) robotic sounds problems for other channel leg occurs (sometimes it happens for both legs) !

After research **G722** codec we realize it has unclear **RFC** and this produce many mistake/BUG(s) for usage side. G722 is extremely complex.

It is tested for both,
-- FreeSwitch Version **1.10.3**-release & **1.10.6** version
-- Debian 8 & Debian 10

Unfortunately  same issue occurs !

Here it is sample pcap file
[g722_codec_change.pcap.zip](https://github.com/signalwire/freeswitch/files/6420145/g722_codec_change.pcap.zip)

**Comment by dragos-oancea** [comment]
does it happen for the rest of the call after the codec change or just for a short period right after the codec is changed ? 
FS DEBUG logs would help.

**Comment by melekoktay** [comment]
Yes, it happen for the rest of the call after the codec changes.
In addition, it happens only G722 codec. If call was established for PCMA or PCMU, this issue is not happening. Only G722

I put appropriate FreeSwitch DEBUG logs in attachment.

[g722_codec_robotic_sound_debug_logs.txt](https://github.com/signalwire/freeswitch/files/6421492/g722_codec_robotic_sound_debug_logs.txt)

**Comment by esualli** [comment]
The resample struct may be initialize with the wrong sample rate.

Since for the bridge call, thread 1 go through from A session to session B, and do resample with session B's resample buffer with bridge function. It means thread 1 may touch both A session and B session. So when we set the codec when we change session B's codec, some old codec packet may be still working in thread 1. It means after we  call switch_core_media_set_codec and reset resample in thread 1 and reset the session B's resample buffer. But there may be some packets with old codec/sample rate still in the thread 1 and init the session B's resample buffer. in switch_core_media_set_codec function, we destroy the codec at the beginning of the function. I think we should destroy resample buffer after we set the codec.

The way I fix this problem was like below. I change the code locally, reset the resample buffer again after set the codec and hope it will work for you.
in switch_core_media_set_codec(switch_core_media.c), after 
```
	switch_core_session_set_read_impl(session, a_engine->read_codec.implementation);
	switch_core_session_set_write_impl(session, a_engine->write_codec.implementation);
```
add
```
	if (session->read_resampler) {
		switch_mutex_lock(session->resample_mutex);
		switch_resample_destroy(&session->read_resampler);
		switch_mutex_unlock(session->resample_mutex);
	}
						
	if (session->write_resampler) {
		switch_mutex_lock(session->resample_mutex);
		switch_resample_destroy(&session->write_resampler);
		switch_mutex_unlock(session->resample_mutex);
	}

	if(switch_core_session_get_partner(session, &other_session) == SWITCH_STATUS_SUCCESS) {
		if (other_session) {
			if (other_session->read_resampler) {
				switch_mutex_lock(other_session->resample_mutex);
				switch_resample_destroy(&other_session->read_resampler);
				switch_mutex_unlock(other_session->resample_mutex);
			}

			if (other_session->write_resampler) {
				switch_mutex_lock(other_session->resample_mutex);
				switch_resample_destroy(&other_session->write_resampler);
				switch_mutex_unlock(other_session->resample_mutex);
			}
			switch_core_session_rwunlock(other_session);
		}
	}
```
